uniffi_bindgen/src/scaffolding/templates/CallbackInterfaceTemplate.rs

```diff
@@ -69,31 +71,82 @@ impl r#{{ trait_name }} for {{ trait_impl }} {
     {#- Calling into foreign code. #}
         let callback = {{ foreign_callback_internals }}.get_callback().unwrap();
 
-        let ret_rbuf = unsafe {
+        unsafe {
             // SAFETY:
             // * We're passing in a pointer to an empty buffer.
             //   * Nothing allocated, so nothing to drop.
             // * We expect the callback to write into that a valid allocated instance of a
             //   RustBuffer.
-            // * A positive return value signals success.
             let mut ret_rbuf = uniffi::RustBuffer::new();
             let ret = callback(self.handle, {{ loop.index }}, args_rbuf, &mut ret_rbuf);
+            #[allow(clippy::let_and_return, clippy::let_unit_value)]
             match ret {
-                0 => uniffi::RustBuffer::new(),
-                _ if ret < 0 => panic!("Callback failed"),
-                _ => ret_rbuf
+                1 => {
+                    // 1 indicates success with the return value written to the RustBuffer for
+                    //   non-void calls.
+                    let result = {
+                        {% match meth.return_type() -%}
+                        {%- when Some(return_type) -%}
+                        let vec = ret_rbuf.destroy_into_vec();
+                        let mut ret_buf = vec.as_slice();
+                        {{ return_type|ffi_converter }}::try_read(&mut ret_buf).unwrap()
+                        {%- else %}
+                        uniffi::RustBuffer::destroy(ret_rbuf);
+                        {%- endmatch %}
+                    };
+                    {%- if meth.throws() %}
+                    Ok(result)
+                    {%- else %}
+                    result
+                    {%- endif %}
+                }
+                -2 => {
+                    // -2 indicates an error written to the RustBuffer
+                    {% match meth.throws_type() -%}
+                    {% when Some(error_type) -%}
+                    let vec = ret_rbuf.destroy_into_vec();
+                    let mut ret_buf = vec.as_slice();
+                    Err({{ error_type|ffi_converter }}::try_read(&mut ret_buf).unwrap())
+                    {%- else -%}
+                    panic!("Callback return -2, but throws_type() is None");
+                    {%- endmatch %}
+                }
+                // 0 is a deprecated method to indicates success for void returns
+                0 => {
```

Comment:
IIUC, we only expect to see this for bindings outside of this crate (and maybe Ruby 😅 ) - so how about arranging for a deprecated warning here? (Deprecating and finally removing things without generating a warning seems very likely to catch people by surprise later)

Reply:
Good call.  I added an `eprintln!` for this, is that the best way to signal a warning?

Reply:
A mildly better way might be to have a noop function marked `#[deprecated] ` and call it here, but I'm really not too bothered